waspc/data/Generator/templates/sdk/wasp/server/operations/wrappers.ts

```diff
@@ -0,0 +1,233 @@
+> = Parameters<OperationDefinition> extends []
+  ? AuthenticatedOperation<void, _Awaited<_ReturnType<OperationDefinition>>>
+  : AuthenticatedOperation<
+      Parameters<OperationDefinition>[0],
+      _Awaited<_ReturnType<OperationDefinition>>
+    >
+
+/**
+ * The type of the context users must pass when calling an authenticated
+ * operation's server-side API.
+ */
+export type AuthenticatedOperationContext = { user: AuthUser }
+
+// PRIVATE API (used in SDK)
+/**
+ * Creates the server-side API for an authenticated operation.
+ *
+ * @template OperationDefinition The type of the authenticated operation's definition.
+ * @param userOperation The authenticated operation's definition.
+ * @param entities The authenticated operation's entity map .
+ * @returns The server-side API for the provided authenticated operation.
+ */
+export function createAuthenticatedOperation<
+  OperationDefinition extends GenericOperationDefinition
+>(
+  userOperation: OperationDefinition,
+  entities: EntityMapFor<OperationDefinition>
+): AuthenticatedOperationFor<OperationDefinition> {
+  async function operation(...args: AuthenticatedOperationArgsFor<OperationDefinition>) {
+    // To understand this function and its limitations, read
+    // https://github.com/wasp-lang/wasp/issues/2050
+    if (args.length < 1) {
+      // No arguments sent -> no user and no payload specified -> there's no way this was called correctly.
+      throw new Error('Invalid number of arguments')
+    } else if (includesPayload(args)) {
+      // Two arguments sent -> the first argument is the payload, the second is the context.
+      const [payload, context] = args
+      return userOperation(payload, {
+        ...context,
+        entities,
+      })
+    } else {
+      // One argument sent -> the first and only argument is the user.
+      const [context] = args
+      return userOperation(undefined as Parameters<OperationDefinition>[0], {
+        ...context,
+        entities,
+      })
+    }
+  }
+
+  return operation as AuthenticatedOperationFor<OperationDefinition>
+}
+
+/**
+ * Returns a boolean carrying compiler type information about whether the
+ * provided arguments array (of an authenticated operation) includes a payload.
+ *
+ * @template Input The type of the payload the operation expects.
+ * @param args The arguments array for an authenticated operation.
+ * @returns true if the arguments array includes a payload, false otherwise.
+ */
+function includesPayload<Input>(
+  args: [AuthenticatedOperationContext] | [Input, AuthenticatedOperationContext]
+): args is [Input, AuthenticatedOperationContext] {
+  return args.length === 2
+}
+
+/**
+ * Constructs the type of the arguments array for an authenticated operation
+ * based on the type of its definition.
+ */
+type AuthenticatedOperationArgsFor<Op extends GenericOperationDefinition> =
+  Parameters<AuthenticatedOperationFor<Op>>
+
+/**
+ * Constructs the type for an authenticated operation's server-side API.
```

Comment:
Server-side API feels okay to me since there are the `wasp/server/operations` and `wasp/client/operations` modules which we are trying to get the users to get used to :D

Reply:
Actually that makes sense, I am not sure why it wasn't clear to me at the moment of the reading. Might be that I got involved in some other context so that threw me off. Sounds good!